### PR TITLE
Added support for typing of monetary amounts

### LINKED ISSUE #46: Type Id Handling not supported
Hi Folks,

I want to serialise Money types with Type Info, so that I can properly use it in serialised collections. In order to achieve this, I use Jackson DefaultTyping for polymorphic serialisation. Otherwise, type information will get lost when I serialise a Collection.

This is my (simplified) scenario:

```java
public class ParentDomainObject {
  private Long id;
  private Set<ChildDomainObject> children;
}

public class ChildDomainObject {
  private Long id;
  private MonetaryAmount cost;
}

public class SomeHandler {
  public void serialiseMyDomainObjects() {
    ObjectMapper mapper = new ObjectMapper();
    mapper.registerModule(new MoneyModule());
    mapper.enableDefaultTyping(ObjectMapper.DefaultTyping.OBJECT_AND_NON_CONCRETE);

    ParentDomainObject o1 = new ParentDomainObject();
    // add some child objects

    ParentDomainObject o2 = new ParentDomainObject();
    // add some child objects

    mapper.writeValueAsString(Sets.asSet(o1, o2));
  }
}
```

When I try to run this, I get the following Exception:

```
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Type id handling not implemented for type org.javamoney.moneta.Money (by serializer of type org.zalando.jackson.datatype.money.MonetaryAmountSerializer)
```

I hope you get the point. Is this simply not supported (at the moment) or is there any way to work around this?

Cheers,
Stefan

### PR DESCRIPTION
**Pull request by whiskeysierra**
Fixes #46

- [x] support deserialization

### CODE REVIEW

**Comment by stefanzilske** [comment]
To me this looks fine, as far a serialisation is concerned. In my example, it now adds the Money type during serialisation, even in nested Collections. Awesome work!!

Support for deserialisation, of course, would be highly appreciated 😉

**Comment by jhorstmann** [comment]
How should deserialization work, use the serialized type info or the configured `TypeFactory`?

Also, the following test with the serialied string taken from `MonetaryAmountSerializerTest` fails because the currency unit is missing type info, so it seems the roundtrip is more tricky.

    @Test
    public void shouldDeserializeWithConfiguredTypeFactory() throws IOException {
        final ObjectMapper unit = unit(module())
                .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
                .enableDefaultTypingAsProperty(OBJECT_AND_NON_CONCRETE, "type");

        final String string = "{\"type\":\"" + FastMoney.class.getName() + "\",\"amount\":29.95,\"currency\":\"EUR\"}";
        final M money = unit.readValue(string, type);

        assertThat(money, instanceOf(type));
    }

The exception message is



    com.fasterxml.jackson.databind.JsonMappingException: Unexpected token (VALUE_STRING), expected FIELD_NAME: missing property 'type' that is to contain type id  (for class javax.money.CurrencyUnit)
        at [Source: {"type":"org.javamoney.moneta.FastMoney","amount":29.95,"currency":"EUR"}; line: 1, column: 68]

**Comment by whiskeysierra** [comment]
@jhorstmann I *solved* it now but essentially not supporting type information for currency values.

**Comment by jhorstmann** [comment]
@whiskeysierra The changes here are mostly for avoiding the exception when a specific feature is enabled, but do you expect a consumer to use this type information? If the consumer is also using this module, they would probably configure a `TypeFactory` for deserialization. So I'm thinking that adding the type to the output is unnecessary overhead.

**Comment by whiskeysierra** [comment]
> The changes here are mostly for avoiding the exception when a specific feature is enabled

Kind of true, yes.

> but do you expect a consumer to use this type information?

Yes, why not?

> If the consumer is also using this module, they would probably configure a TypeFactory for deserialization. 

Yes, they would and it would be taken into account whenever a monetary without type information is being deserialized.

> So I'm thinking that adding the type to the output is unnecessary overhead, and we could treat Money basically as a primitive without type information.

We do that by default, but when somebody enables default typing on purpose we should behave as requested and not silently ignore it.

**Comment by whiskeysierra** [comment]
@stefanzilske Can you take a look at the implementation?

**Comment by whiskeysierra** [comment]
@stefanzilske Do you actually need type information for money/currency or do you just want the module to work in conjunction with default typing?

**Comment by stefanzilske** [comment]
Hi folks, sorry for being so silent at the moment. Will check the implementation asap.

@whiskeysierra, not sure if I get your question. As mentioned in #46, I need to serialize nested collections, of domain objects, which again include money types. To achieve this, I need to enable the DefaultTyping, which did not work with the module. Please let me know if I need to provide more information!

**Comment by whiskeysierra** [comment]
> As mentioned in #46, I need to serialize nested collections, of domain objects, which again include money types. To achieve this, I need to enable the DefaultTyping

That is not true, actually. For that particular example you don't need type information. In fact you never need them during serialization. If at all you'll need it during deserialization (not in your example though).

*But*, since Jackson does come with that feature it would definitely be nice if our module would just work.

Now we have two different options how to support typing:
1. Actually properly support, i.e. we add type information
2. We somewhat support, i.e. we don't add type information but we don't fail either

1 is the go-to solution for most data types while 2 is Jackson's way to handle what they call *natural, scalar* values, e.g. String, int, boolean, ...

Although money is an object we'd still consider it to be a scalar value and therefore tend to prefer option 2.

My question now is, which one do you prefer?

File: README.md [hunk 1]
```diff
@@ -91,7 +91,7 @@ ObjectMapper mapper = new ObjectMapper()
 The `DefaultMonetaryAmountFormatFactory` delegates directly to `MonetaryFormats.getAmountFormat(Locale, String...)`.
 
 Formatting only affects the serialization and can be customized based on the *current* locale, as defined by the
-[`SerializationConfig`](https://fasterxml.github.io/jackson-databind/javadoc/2.0.0/com/fasterxml/jackson/databind/SerializationConfig.html#with\(java.util.Locale\)). This allows to implement RESTful API endpoints
+[`SerializationConfig`](http://wiki.fasterxml.com/SerializationConfig). This allows to implement RESTful API endpoints
```
Comment (AlexanderYastrebov) [inline README.md]: Looks like it is breaking link again

Comment (whiskeysierra) [inline README.md]: That's weird

File: src/test/java/org/zalando/jackson/datatype/money/MonetaryAmountSerializerTest.java [hunk 2]
```diff
@@ -151,4 +152,19 @@ public void shouldRespectJacksonFeatureByDefault() throws JsonProcessingExceptio
         assertThat(actual, is(expected));
     }
 
+    @Value
+    private static final class Price {
+        MonetaryAmount amount;
+    }
+
+    @Test
+    public void shouldSerializeWithType() throws JsonProcessingException {
```
Comment (AlexanderYastrebov) [inline src/test/java/org/zalando/jackson/datatype/money/MonetaryAmountSerializerTest.java]: WithoutType

**Comment by AlexanderYastrebov** [comment]
:+1:

**Pull request merged**